internal/controller/configmap.go

```diff
@@ -0,0 +1,419 @@
+	}
+	return nil
+}
+
+// reconcileRBACforBuiltInRole will ensure that the ArgoCD RBAC ConfigMap is present for the built-in roles.
+func (r *ArgoCDRoleBindingReconciler) reconcileRBACForBuiltInRole(rb *argoprojiov1alpha1.ArgoCDRoleBinding) error {
+	cm := newConfigMap()
+	role := &argoprojiov1alpha1.ArgoCDRole{}
+
+	switch rb.Spec.ArgoCDRoleRef.Name {
+	case "admin":
+		role = createBuiltInAdminRole()
+	case "readonly":
+		role = createBuiltInReadOnlyRole()
+	}
+
+	if IsObjectFound(r.Client, cm.Namespace, cm.Name, cm) {
+		return r.reconcileRBACConfigMapForBuiltInRole(cm, rb, role)
+	}
+	return r.createRBACConfigMapForBuiltInRole(cm, rb, role)
+}
+
+// reconcileRBACConfigMap will ensure that the ArgoCD RBAC ConfigMap is up-to-date.
+func (r *ArgoCDRoleBindingReconciler) reconcileRBACConfigMapForBuiltInRole(cm *corev1.ConfigMap, rb *argoprojiov1alpha1.ArgoCDRoleBinding, role *argoprojiov1alpha1.ArgoCDRole) error {
+	changed := false
+	overlayKey := fmt.Sprintf("policy.%s.csv", role.Name)
+
+	// Default Policy String
+	if cm.Data[common.ArgoCDKeyRBACPolicyCSV] != getDefaultRBACPolicy() {
+		cm.Data[common.ArgoCDKeyRBACPolicyCSV] = getDefaultRBACPolicy()
+		changed = true
+	}
+	// Policy OverlayKey CSV
+	if cm.Data[overlayKey] != buildPolicyStringSubjects(rb, role) {
+		cm.Data[overlayKey] = buildPolicyStringSubjects(rb, role)
+		changed = true
+	}
+
+	if changed {
+		return r.Client.Update(context.TODO(), cm)
+	}
+	return nil
+}
+
+// createRBACConfigMap will create the Argo CD RBAC ConfigMap resource.
+func (r *ArgoCDRoleBindingReconciler) createRBACConfigMapForBuiltInRole(cm *corev1.ConfigMap, rb *argoprojiov1alpha1.ArgoCDRoleBinding, role *argoprojiov1alpha1.ArgoCDRole) error {
+	data := make(map[string]string)
+	overlayKey := fmt.Sprintf("policy.%s.csv", role.Name)
+
+	data[common.ArgoCDKeyRBACPolicyDefault] = getRBACPolicyDefault()
+	data[common.ArgoCDKeyRBACScopes] = getRBACScopes()
+	data[common.ArgoCDKeyRBACPolicyCSV] = getDefaultRBACPolicy()
+	data[overlayKey] = buildPolicyStringSubjects(rb, role)
+	cm.Data = data
+
+	return r.Client.Create(context.TODO(), cm)
+}
+
+// IsObjectFound will perform a basic check that the given object exists via the Kubernetes API.
+// If an error occurs as part of the check, the function will return false.
+func IsObjectFound(client client.Client, namespace string, name string, obj client.Object) bool {
+	return !apierrors.IsNotFound(FetchObject(client, namespace, name, obj))
+}
+
+// FetchObject will retrieve the object with the given namespace and name using the Kubernetes API.
+// The result will be stored in the given object.
+func FetchObject(client client.Client, namespace string, name string, obj client.Object) error {
+	return client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, obj)
+}
+
+// createBuiltInAdminRole will return a new built-in ArgoCDRole with admin permissions.
+func createBuiltInAdminRole() *argoprojiov1alpha1.ArgoCDRole {
```

Comment:
There role already exist in the argo codebase. Why duplicating them?

Reply:
the duplicates are merely for localUser handling, since the policies must be applied to the localUser directly and not through "g, localUser, role:admin"